dashboard/app/models/pd/workshop.rb

```diff
@@ -306,6 +306,26 @@ def end!
     return unless ended_at.nil?
     self.ended_at = Time.zone.now
     save!
+
+    # We want to send exit surveys now, but that needs to be done on the
+    # production-daemon machine, so we'll let the process_pd_workshop_emails
+    # cron job call the process_ends function below on that machine.
+  end
+
+  # This is called by the process_pd_workshop_emails cron job which is run
+  # on the production-daemon machine, and will send exit surveys to workshops
+  # that have been ended in the last two days when they haven't already had
+  # that done.
+  # The emails must be sent from production-daemon because they contain attachments.
+  # See https://github.com/code-dot-org/code-dot-org/blob/96b890d6e6f77de23bc5d4469df69b900e3fbeb7/lib/cdo/poste.rb#L217
+  # for details.
+  def self.process_ends
+    end_on_or_after(Time.now - 2.days).each do |workshop|
+      if !workshop.processed_at || workshop.processed_at < workshop.ended_at
```

Comment:
It'd be nice to have a test demonstrating that a workshop that is closed, opened, closed again will send emails the second time.

Reply:
Good point, and shall do, though my understanding is that while we will attempt to send surveys again, the safeguard in the enrollment code will prevent it from actually sending them again.

Reply:
I don't think we do re-send emails the second time a workshop is closed, unless we added a participant who didn't receive the email the first time

Reply:
after re-reading the description, it sounds like this behavior hasn't changed, and a test that it will send any surveys that weren't already sent would be good.

Reply:
ha i should have refreshed before commenting....